### PR TITLE
Refactor Tokens contract

### PR DESCRIPTION
**Pull request by jmjatlanta**
NOTE 1: This builds on the jmj_ccinit branch. It will make for easier review if PR #444 was reviewed first.
NOTE 2: There have undoubtedly been a number of logic changes to the tokens contract as part of the Tokel project, Merging this into those changes will probably be difficult. If the ideas in this PR are desirable, it may be best to replay these commits after the Tokel changes are merged in (or redo this PR manually).

The functionality of the CC Tokens code was spread within several files. This PR moves them into a few, and wraps then into a class. In some cases, this eliminated the need to pass the CCcontract_info object as a parameter. In other cases, this required the call to static methods within the class. 

NOTE 3: the function CCTokens::GetCCaddress is an example where converting this from a pseudo-override type function of (::GetCCaddress) to a template would avoid the runtime hit of indirection as well as increase type safety by removal of a few dynamic_casts, in exchange for increased compile time. ::GetCCaddress may be a candidate for something that should be within the CCcontract_info class, but that has not been explored.

### CODE REVIEW

File: src/cc/CCcontract.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include "pubkey.h"
+#include "script/script.h"
+
+/// CC contract (Antara module) info structure that contains data used for signing and validation of cc contract transactions
+struct CCcontract_info
+{
+    /***
+     * Default ctor, sets everything to 0
+     */
+    CCcontract_info() { memset(this, 0, sizeof(CCcontract_info)); }
```
Comment (DeckerSU) [inline src/cc/CCcontract.h:13]: Here the `memset` function is used to nullify fields of `CCcontract_info` class. Virtual table pointer (vptr) could be damaged, bcz `memset` doesn't know anything about vptr and `sizeof(CCcontract_info)` is virtual methods existence dependent. May be such default ctor with member initializer list will be more correct?
```
struct CCcontract_info
{

    uint8_t evalcode;
    uint8_t additionalTokensEvalcode2 = 0;
    char unspendableCCaddr[64];
    // < members skipped >
    CCcontract_info():   evalcode(0), additionalTokensEvalcode2(0),
                         unspendableCCaddr{0} /*, etc. */
                         {}
    virtual ~CCcontract_info() {}
    // < virtual methods skipped >
};
```

Comment (dimxy) [reply src/cc/CCcontract.h:13]: yes i'd prefer to avoid using memset with c++ objects, as it may override members' constructor initialisations

**Comment by dimxy** [comment]
found one catch
there is a MakeCC1vout function in CCutils.cpp
and in tokens there was an additional function MakeTokensCC1vout (which does the same always add token evalcode). Now the MakeTokensCC1vout func is renamed also to MakeCC1vout.
in CCTokens::CreateToken() they are called both. But now the global MakeCC1vout (intended to be called) is obscured by CCTokens::MakeCC1vout, so now we need to call it with ::MakeCC1vout
That is, MakeTokensCC1vout was not planned to be an overload but an additional function

**Comment by jmjatlanta** [comment]
> But now the global MakeCC1vout (intended to be called) is obscured by CCTokens::MakeCC1vout, so now we need to call it with ::MakeCC1vout That is, MakeTokensCC1vout was not planned to be an overload but an additional function

I thought I took care of that, but I'll research all of these issues. Thanks @DeckerSU  and @dimxy for the reviews.

**Comment by jmjatlanta** [comment]
I was certainly calling the incorrect method for MakeCC1vout. I believe that has been fixed. I will also attempt to build a test to prove that it is fixed.

As for the vtable issue, I was certainly stepping on the pointer. Fixing this requires a sizable adjustment. As currently implemented, CCcontract_info needs to be in the common lib in order for the virtual method implementations to link correctly. That comes with a lot of unintended side effects that I want to avoid. I am taking a step back to see what other options are available.

**Comment by dimxy** [comment]
This PR should be retargeted to the 'research' branch (where cc code is developed first)

**Pull request closed without merging**